otel-integration/k8s-helm/e2e-test/sinks.go

```diff
@@ -0,0 +1,198 @@
+package e2e
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+	"slices"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer/consumertest"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	"go.opentelemetry.io/collector/receiver/receivertest"
+)
+
+type ExpectedValueMode int
+
+const (
+	AttributeMatchTypeEqual ExpectedValueMode = iota
+	AttributeMatchTypeRegex
+	AttributeMatchTypeExist
+	UidRe = "[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}"
+
+	ServiceNameAttribute = "service.name"
+)
+
+type ExpectedValue struct {
+	Mode  ExpectedValueMode
+	Value string
+}
+
+func NewExpectedValue(mode ExpectedValueMode, value string) ExpectedValue {
+	return ExpectedValue{
+		Mode:  mode,
+		Value: value,
+	}
+}
+
+func StartUpSinks(t *testing.T, mc *consumertest.MetricsSink, tc *consumertest.TracesSink) func() {
+	f := otlpreceiver.NewFactory()
+	cfg := f.CreateDefaultConfig().(*otlpreceiver.Config)
+	cfg.HTTP = nil
+	cfg.GRPC.NetAddr.Endpoint = "0.0.0.0:4317"
+
+	_, err := f.CreateMetrics(context.Background(), receivertest.NewNopSettings(), cfg, mc)
+	require.NoError(t, err, "failed creating metrics receiver")
+	rcvr, err := f.CreateTraces(context.Background(), receivertest.NewNopSettings(), cfg, tc)
+	require.NoError(t, err, "failed creating traces receiver")
+	require.NoError(t, rcvr.Start(context.Background(), componenttest.NewNopHost()))
+
+	return func() {
+		assert.NoError(t, rcvr.Shutdown(context.Background()))
+	}
+}
+
+func WaitForMetrics(t *testing.T, entriesNum int, mc *consumertest.MetricsSink) {
+	timeoutSeconds := 180 // 3 minutes
+	require.Eventuallyf(t, func() bool {
+		count := len(mc.AllMetrics())
+		t.Logf("Waiting for metrics: got %d/%d", count, entriesNum)
+		return count >= entriesNum
+	}, time.Duration(timeoutSeconds)*time.Second, 1*time.Second,
+		"failed to receive %d entries in %d seconds",
+		entriesNum, timeoutSeconds)
+}
+
+func WaitForTraces(t *testing.T, entriesNum int, tc *consumertest.TracesSink) {
+	timeoutSeconds := 180 // 3 minutes
+	require.Eventuallyf(t, func() bool {
+		count := len(tc.AllTraces())
+		t.Logf("Waiting for traces: got %d/%d", count, entriesNum)
+		return count >= entriesNum // Changed > to >=
+	}, time.Duration(timeoutSeconds)*time.Second, 1*time.Second,
+		"failed to receive %d entries in %d seconds",
+		entriesNum, timeoutSeconds)
+}
+
+func ScanTracesForAttributes(t *testing.T, ts *consumertest.TracesSink, expectedService string, kvs map[string]ExpectedValue, scopeSpanAttrs []map[string]ExpectedValue) {
+	for i := 0; i < len(ts.AllTraces()); i++ {
+		traces := ts.AllTraces()[i]
+		for i := 0; i < traces.ResourceSpans().Len(); i++ {
```

Comment:
The two loops use the same variable for control. This won't behave well. I see that  `ts.AllTraces()` is an array, so the normal `for-range` loop can be used. The same does not apply for `traces.ResourceSpans()`, so that C-style for loop has to stay for this one.

Reply:
removed this function, so would be irrelevant